// tmplconv/shim/node/converting/convert_dir.js
/**
 * Convert files in a directory.
 * @function convertDir
 * @param {string} srcDir - Source directory path.
 * @param {string} destDir - Destination directory path.
 * @param {object} [options] - Optional settings.
 * @param {string|string[]} [options.pattern] - Filename pattern.
 * @param {string|string[]} [options.ignore] - Filename pattern.
 * @param {object} [options.rule] - Convert map.
 * @param {string} [options.mode='644'] - File permission to generate.
 * @param {boolean} [options.clean=false] - Cleanup destination directory before convert.
 * @param {boolean} [options.once=false] - Write only first time. Skip if already exists.
 * @param {function} [options.out] - Convert output file name.
 * @returns {Promise}
 */
'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var co = require('co');
var argx = require('argx');
var path = require('path');

var _require = require('asfs');

var existsAsync = _require.existsAsync;

var rimraf = require('rimraf');
var aglob = require('aglob');
var convertString = require('./convert_string');

/** @lends convertDir */
function convertDir(srcDir, destDir) {
  var options = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

  var args = argx(arguments);
  options = args.pop('object');

  var pattern = options.pattern || '**/*.*';
  var ignore = [].concat(options.ignore || []);
  var rule = options.rule || {};
  var out = options.out;
  var once = !!options.once;
  var clean = !!options.clean;

  return co(_regenerator2.default.mark(function _callee() {
    var exists, filenames, results, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, filename, src, dest, result;

    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return existsAsync(srcDir);

          case 2:
            exists = _context.sent;

            if (exists) {
              _context.next = 5;
              break;
            }

            throw new Error('srcDir not exists: ' + srcDir);

          case 5:
            if (!clean) {
              _context.next = 8;
              break;
            }

            _context.next = 8;
            return new _promise2.default(function (resolve, reject) {
              return rimraf(destDir, function (err) {
                return err ? reject(err) : resolve();
              });
            });

          case 8:
            _context.next = 10;
            return aglob(pattern, {
              cwd: srcDir,
              ignore: ignore
            });

          case 10:
            filenames = _context.sent;

            filenames = filenames.filter(function (filename, i, filenames) {
              return filenames.indexOf(filename) === i;
            });
            results = [];
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            _context.prev = 16;
            _iterator = (0, _getIterator3.default)(filenames);

          case 18:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              _context.next = 30;
              break;
            }

            filename = _step.value;
            src = path.resolve(srcDir, filename);
            dest = path.resolve(destDir, convertString(filename, rule));

            if (out) {
              dest = out(dest);
            }
            _context.next = 25;
            return require('./convert')(src, dest, {
              rule: rule,
              once: once,
              mode: options.mode
            });

          case 25:
            result = _context.sent;

            results.push(result);

          case 27:
            _iteratorNormalCompletion = true;
            _context.next = 18;
            break;

          case 30:
            _context.next = 36;
            break;

          case 32:
            _context.prev = 32;
            _context.t0 = _context['catch'](16);
            _didIteratorError = true;
            _iteratorError = _context.t0;

          case 36:
            _context.prev = 36;
            _context.prev = 37;

            if (!_iteratorNormalCompletion && _iterator.return) {
              _iterator.return();
            }

          case 39:
            _context.prev = 39;

            if (!_didIteratorError) {
              _context.next = 42;
              break;
            }

            throw _iteratorError;

          case 42:
            return _context.finish(39);

          case 43:
            return _context.finish(36);

          case 44:
            return _context.abrupt('return', results);

          case 45:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[16, 32, 36, 44], [37,, 39, 43]]);
  }));
}

module.exports = convertDir;
