// tmplconv/shim/node/render.js
/**
 * Render template.
 * @memberof module:tmplconv/lib
 * @function render
 * @param {string} srcDir - Name of source directory, which contains template files.
 * @param {string} destDir - Name of destination directory.
 * @param {object} options - Optional settings.
 * @param {string|object} options.data - Data for tmpls.
 * @param {string|string[]} options.pattern - Source patterns.
 * @param {string|string[]} [options.ignore] - Filename pattern.
 * @param {string} [options.prefix='_____'] - Embed prefix.
 * @param {string} [options.suffix='_____'] - Embed suffix.
 * @param {string} [options.extname='.tmpl'] - Template extension name.
 * @param {boolean} [options.silent=false] - Silent or not.
 * @param {string} [options.mode='644'] - File permission to generate.
 * @param {boolean} [options.clean=false] - Cleanup destination directory before convert.
 * @param {boolean} [options.once=false] - Write only first time. Skip if already exists.
 * @returns {Promise}
 */

'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var argx = require('argx');
var co = require('co');
var path = require('path');
var convert = require('./converting/convert');
var _rule = require('./_rule');
var removeExtname = require('./naming/remove_extname');
var _logResults = require('./_log_results');

/** @lends render */
function render(srcDir, destDir, options) {
  var args = argx(arguments);
  if (args.pop('function')) {
    throw new Error('Callback is no longer supported. Use promise interface instead.');
  }
  options = args.pop('object') || {};

  return co(_regenerator2.default.mark(function _callee() {
    var results, shouldLog;
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return convert(srcDir, destDir, {
              pattern: options.pattern || '**/*.*',
              rule: _rule(options.data, options.prefix, options.suffix),
              ignore: options.ignore,
              mode: options.mode,
              clean: options.clean,
              once: options.once,
              out: function out(src) {
                return removeExtname(src, options.extname || '.tmpl');
              }
            });

          case 2:
            results = _context.sent;
            shouldLog = !options.silent;

            if (shouldLog) {
              _logResults(results);
            }

            // Wait for flush
            _context.next = 7;
            return new _promise2.default(function (resolve) {
              return process.nextTick(function () {
                return resolve();
              });
            });

          case 7:
            return _context.abrupt('return', results);

          case 8:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));
}

module.exports = render;
