// tmplconv/shim/node/transplant.js
/**
 * Tmplify and render at once.
 * @memberof module:tmplconv/lib
 * @function transplant
 * @param {string} src - Name of destination directory.
 * @param {string} dest - Name of destination directory, which contains template files.
 * @param {object} options - Optional settings.
 * @param {string|object} options.rule - Rule for convert.
 * @param {string|string[]} options.pattern - Source patterns.
 * @param {string|string[]} [options.ignore] - Filename pattern.
 * @param {boolean} [options.silent=false] - Silent or not.
 * @param {string} [options.mode='644'] - File permission to generate.
 * @param {boolean} [options.clean=false] - Cleanup destination directory before convert.
 * @param {boolean} [options.once=false] - Write only first time. Skip if already exists.
 * @returns {Promise}
 */

'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var argx = require('argx');
var co = require('co');
var path = require('path');
var tmplify = require('./tmplify');
var rimraf = require('rimraf');
var render = require('./render');

/** @lends transplant */
function transplant(src, dest, options) {
  var args = argx(arguments);
  if (args.pop('function')) {
    throw new Error('Callback is no longer supported. Use promise interface instead.');
  }
  options = args.pop('object') || {};

  var tmp = _nameTmp(dest);
  var tmplifyDo = {},
      renderDo = {};
  var rule = options.rule || {};
  (0, _keys2.default)(rule).forEach(function (src, i) {
    var key = 'key_' + i;
    tmplifyDo[key] = src;
    renderDo[key] = rule[src];
  });
  return co(_regenerator2.default.mark(function _callee() {
    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return new _promise2.default(function (resolve, reject) {
              return rimraf(tmp, function (err) {
                return err ? reject(err) : resolve();
              });
            });

          case 2:
            _context.next = 4;
            return tmplify(src, tmp, {
              data: tmplifyDo,
              silent: true,
              pattern: options.pattern,
              ignore: options.ignore
            });

          case 4:
            _context.next = 6;
            return render(tmp, dest, {
              data: renderDo,
              silent: options.silent,
              clean: options.clean,
              once: options.once,
              mode: options.mode
            });

          case 6:
            _context.next = 8;
            return new _promise2.default(function (resolve, reject) {
              return rimraf(tmp, function (err) {
                return err ? reject(err) : resolve();
              });
            });

          case 8:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));
}

function _nameTmp(dest) {
  var dirname = path.dirname(dest);
  var basename = path.basename(dest);
  return path.join(dirname, '.' + basename + '.tmp');
}

module.exports = transplant;
