// asfs/shim/node/asfs.js
/**
 * @module asfs
 */
'use strict';

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var fs = require('fs');
var mkdirp = require('mkdirp');

module.exports = (0, _assign2.default)(exports, {
  /**
   * Check file exists
   * @param {string} filename
   * @returns {Promise.<boolean>}
   */
  existsAsync: function existsAsync(filename) {
    return new _promise2.default(function (resolve) {
      return fs.exists(filename, function (exists) {
        return resolve(exists);
      });
    });
  },


  /**
   * Make directories with parent option.
   * @param {string} dirname
   * @returns {Promise}
   */
  mkdirpAsync: function mkdirpAsync(dirname) {
    return new _promise2.default(function (resolve, reject) {
      return mkdirp(dirname, function (err) {
        return err ? reject(err) : resolve();
      });
    });
  },


  /**
   * Read file
   * @param {string} filename
   * @param {string} [encode]
   * @returns {Promise}
   */
  readFileAsync: function readFileAsync(filename, encode) {
    var args = [].concat(Array.prototype.slice.call(arguments));
    return new _promise2.default(function (resolve, reject) {
      return fs.readFile.apply(fs, (0, _toConsumableArray3.default)(args.concat(function (err, content) {
        return err ? reject(err) : resolve(content);
      })));
    });
  },


  /**
   * Write a file
   * @param {string} filename
   * @param {string} content
   * @returns {Promise}
   */
  writeFileAsync: function writeFileAsync(filename, content) {
    var args = [].concat(Array.prototype.slice.call(arguments));
    return new _promise2.default(function (resolve, reject) {
      return fs.writeFile.apply(fs, (0, _toConsumableArray3.default)(args.concat(function (err) {
        return err ? reject(err) : resolve();
      })));
    });
  },


  /**
   * Unlink a file
   * @param {string} filename
   * @returns {Promise}
   */
  unlinkAsync: function unlinkAsync(filename) {
    return new _promise2.default(function (resolve, reject) {
      return fs.unlink(filename, function (err) {
        return err ? reject(err) : resolve();
      });
    });
  },


  /**
   * Get state of a file
   * @param {string} filename
   * @returns {Promise}
   */
  statAsync: function statAsync(filename) {
    return new _promise2.default(function (resolve, reject) {
      return fs.stat(filename, function (err, stat) {
        return err ? reject(err) : resolve(stat);
      });
    });
  },


  /**
   * Read directory
   * @param {string} dirname
   * @param {Object} [options]
   * @returns {Promise}
   */
  readdirAsync: function readdirAsync(dirname) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    return new _promise2.default(function (resolve, reject) {
      return fs.readdir(dirname, options, function (err, result) {
        return err ? reject(err) : resolve(result);
      });
    });
  }
});
